crates/sui-core/src/authority.rs

```diff
@@ -4649,6 +4654,97 @@ impl AuthorityState {
     }
 }
 
+pub struct RandomnessRoundReceiver {
+    authority_state: Arc<AuthorityState>,
+    randomness_rx: mpsc::Receiver<(EpochId, RandomnessRound, Vec<u8>)>,
+}
+
+impl RandomnessRoundReceiver {
+    pub fn spawn(
+        authority_state: Arc<AuthorityState>,
+        randomness_rx: mpsc::Receiver<(EpochId, RandomnessRound, Vec<u8>)>,
+    ) -> JoinHandle<()> {
+        let rrr = RandomnessRoundReceiver {
+            authority_state,
+            randomness_rx,
+        };
+        spawn_monitored_task!(rrr.run())
+    }
+
+    async fn run(mut self) {
+        info!("RandomnessRoundReceiver event loop started");
+
+        loop {
+            tokio::select! {
+                maybe_recv = self.randomness_rx.recv() => {
+                    if let Some((epoch, round, bytes)) = maybe_recv {
+                        self.handle_new_randomness(epoch, round, bytes);
+                    } else {
+                        break;
+                    }
+                },
+            }
+        }
+
+        panic!("RandomnessRoundReceiver event loop should never end");
+    }
+
+    #[instrument(level = "debug", skip_all, fields(?epoch, ?round))]
+    fn handle_new_randomness(&self, epoch: EpochId, round: RandomnessRound, bytes: Vec<u8>) {
+        let epoch_store = self.authority_state.load_epoch_store_one_call_per_task();
+        if epoch_store.epoch() != epoch {
+            error!(
+                "BUG: dropping randomness for epoch {epoch}, round {round}, because we are in epoch {}",
```

Comment:
Curious, why is randomness network loop is not per-epoch? Is it an optimization to allow it receive future epoch messages when it's behind?

Reply:
That is one reason, and also it would take a nontrivial amount of extra work to make the networking stuff work per-epoch. Network init (including routing setup etc) happens once at startup, the validators don't all reconnect at the start of each epoch. In any case there didn't seem to be much downside to keeping the network loop running persistently